static/app/utils/profiling/renderers/textRenderer.tsx

```diff
@@ -0,0 +1,164 @@
+import {mat3} from 'gl-matrix';
+
+import {Flamegraph} from '../flamegraph';
+import {FlamegraphTheme} from '../flamegraph/FlamegraphTheme';
+import {FlamegraphFrame} from '../flamegraphFrame';
+import {
+  ELLIPSIS,
+  findRangeBinarySearch,
+  getContext,
+  Rect,
+  resizeCanvasToDisplaySize,
+  trimTextCenter,
+} from '../gl/utils';
+
+export function isOutsideView(frame: Rect, view: Rect, inverted: boolean): boolean {
+  // Frame is outside of the view on the left
+  if (frame.right <= view.left) {
+    return true;
+  }
+
+  // Frame is outside of the view on the right
+  if (frame.left >= view.right) {
+    return true;
+  }
+
+  // Frame is above the view
+  if (frame.bottom <= view.top) {
+    return true;
+  }
+
+  // Frame is below the view
+  if (frame.top >= view.bottom) {
+    return true;
+  }
+
+  // @TODO check if we still need this
+  if (inverted) {
+    if (frame.top - 1 >= view.bottom) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+class TextRenderer {
+  textCache: Record<string, number> = {};
+
+  canvas: HTMLCanvasElement;
+  context: CanvasRenderingContext2D;
+  theme: FlamegraphTheme;
+  flamegraph: Flamegraph;
+
+  constructor(canvas: HTMLCanvasElement, flamegraph: Flamegraph, theme: FlamegraphTheme) {
+    this.canvas = canvas;
+    this.theme = theme;
+    this.flamegraph = flamegraph;
+
+    this.context = getContext(canvas, '2d');
+
+    resizeCanvasToDisplaySize(canvas);
+  }
+
+  clearCache(): void {
+    this.textCache = {};
+  }
+
+  measureText(context: CanvasRenderingContext2D, text: string): number {
+    if (this.textCache[text]) {
+      return this.textCache[text];
+    }
+    this.textCache[text] = context.measureText(text).width;
+    return this.textCache[text];
+  }
+
+  draw(configViewSpace: Rect, configSpace: Rect, configToPhysicalSpace: mat3): void {
+    this.context.font = `${
+      this.theme.SIZES.BAR_FONT_SIZE * window.devicePixelRatio
+    }px "Source Code Pro", Courier, monospace`;
+
+    this.context.textBaseline = 'alphabetic';
+    this.context.fillStyle = this.theme.COLORS.LABEL_FONT_COLOR;
+
+    const minWidth = this.measureText(this.context, `${ELLIPSIS}`);
+
+    let frame: FlamegraphFrame;
+    let i: number;
+    const length: number = this.flamegraph.frames.length;
+
+    // We currently iterate over all frames, but we could optimize this to only iterate over visible frames.
+    // This could be achieved by querying the flamegraph tree (as an interval tree). This would still run in O(n), but
+    // would improve our best case performance (e.g. when users) zoom into the flamegraph
+    for (i = 0; i < length; i++) {
+      frame = this.flamegraph.frames[i];
```

Comment:
Nit:
```suggestion
    for (const frame of this.flamegraph.frames) {
```

Reply:
Actually, for of loops are notoriously slow (I presume because they use generators), and we iterate over enough frames here that it actually makes a difference. Maybe in the future we can draw all of this using offscreen canvases :)